lib/assert.js

```diff
@@ -341,6 +358,197 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
   }
 };
 
+function isPlainObject(obj) {
+  if (typeof obj !== 'object' || obj === null) {
+    return false;
+  }
+  const proto = ObjectGetPrototypeOf(obj);
+  return proto === ObjectPrototype || proto === null || ObjectPrototypeToString(obj) === '[object Object]';
+}
+
+/**
+ * Compares two objects or values recursively to check if they are equal.
+ * @param {any} actual - The actual value to compare.
+ * @param {any} expected - The expected value to compare.
+ * @param {boolean} [loose=false] - Whether to use loose comparison (==) or strict comparison (===). Defaults to false.
+ * @param {Set} [comparedObjects=new Set()] - Set to track compared objects for handling circular references.
+ * @returns {boolean} - Returns `true` if the actual value matches the expected value, otherwise `false`.
+ * @example
+ * // Loose comparison (default)
+ * compareBranch({a: 1, b: 2, c: 3}, {a: 1, b: '2'}); // true
+ *
+ * // Strict comparison
+ * compareBranch({a: 1, b: 2, c: 3}, {a: 1, b: 2}, true); // true
+ */
+function compareBranch(
+  actual,
+  expected,
+  loose = false,
+  comparedObjects = new SafeSet(),
+) {
+  // Check for Map object equality
+  if (isMap(actual) && isMap(expected)) {
+    if (actual.size !== expected.size) {
+      return false;
+    }
+    const safeIterator = FunctionPrototypeCall(SafeMap.prototype[SymbolIterator], actual);
+
+    for (const { 0: key, 1: val } of safeIterator) {
+      if (!MapPrototypeHas(expected, key)) {
+        return false;
+      }
+      if (!compareBranch(val, MapPrototypeGet(expected, key), loose, comparedObjects)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Check for Set object equality
+  if (isSet(actual) && isSet(expected)) {
+    if (actual.size !== expected.size) {
+      return false;
+    }
+    const safeIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual);
+
+    for (const item of safeIterator) {
+      if (!SetPrototypeHas(expected, item)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Check if expected array is a subset of actual array
+  if (ArrayIsArray(actual) && ArrayIsArray(expected)) {
+    if (expected.length > actual.length) {
+      return false;
+    }
+    if (typeof actual === 'string') {
+      return StringPrototypeIncludes(actual, expected);
+    }
+
+    if (loose) {
+      return ArrayPrototypeEvery(expected, (item) => ArrayPrototypeSome(actual, (actualItem) => {
+        // eslint-disable-next-line eqeqeq
+        return actualItem == item;
+      }));
+    }
+
+    return ArrayPrototypeEvery(expected, (item) => ArrayPrototypeIncludes(actual, item));
+  }
+
+  // Check non object types equality
+  if (!isPlainObject(actual) || !isPlainObject(expected)) {
```

Comment:
I don't understand what "non object types" mean, could you elaborate? I find it a bit weird that we'd need a `isPlainObject` when (AFAIK) we have needed it before, I'm not sure it would be a good precedent – it seems to me all objects should be treated the same.

Reply:
This is a comment left from the OP of the MR that , tbh, confused me a bit as well when I first read it.

The block is supposed to execute whenever you are comparing types where `actual` AND `expected` are not objects

Reply:
If we're dealing with non-objects, the check should be `typeof … !== 'object'`, not that `isPlainObject` object (which is not documented and would be very hard to document)

Reply:
@aduh95 it is a lot more complicated than that...

types that would be considered objects:

1. regexes
2. Dates
3. Errors
4. nulls

and the list is much longer than I can possibly remember 😄 that function is just checking that it is going in the if block if one of those cases happens

I could separate all those cases, for readability, but it would be just for that

Reply:
It's not just readability, it's also about being consistent. `isPlainObject` is an anti-pattern IMO, if we want to e.g. special case regexes, let's do that explicitly.

Reply:
but we don't want to add special cases for e.g. regexes, we want the opposite, we only want to treat differently the case where you don't have 2 plain objects to compare, which is what the if statement is doing

Reply:
I'd argue there's no way to detect when an object is a plain object vs a special object – except on the case where it's a specific "other kind" of object (the ones that have a specific internal slot defined in the spec, e.g. arrays, regexes, proxies, etc.) – and so we should not try to define arbitrary conditions for what constitutes a plain object.

Reply:
I am getting a little confused then, what is the implementation you are suggesting?